scripts/gha/desktop_tester.py

```diff
@@ -0,0 +1,129 @@
+
+r"""Runs and validates Unity standalone testapps.
+
+Usage:
+  python desktop_tester.py --testapp_dir ~/Downloads/testapps
+
+This will search --testapp_dir for integration testapps whose name is given by
+--testapp_name (default 'testapp').
+
+"""
+
+import os
+import platform
+import subprocess
+import threading
+import time
+
+from absl import app
+from absl import flags
+from absl import logging
+import attr
+
+from integration_testing import test_validation
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string(
+    "testapp_dir", None, "Look for testapps recursively in this directory.")
+flags.DEFINE_string(
+    "testapp_name", "testapp",
+    "Name of the testapps. Behaviour differs on MacOS and Windows. For"
+    " 'app', this will look for a file named 'app' on Linux, an 'app.app'"
+    " directory on Mac, and a file named 'app.exe' on Windows.")
+
+
+def main(argv):
+  if len(argv) > 1:
+    raise app.UsageError("Too many command-line arguments.")
+
+  testapp_dir = _fix_path(FLAGS.testapp_dir)
+  testapp_name = FLAGS.testapp_name
+
+  if platform.system() == "Windows":
+    testapp_name += ".exe"
+
+  logging.info("Searching for file named '%s' in %s", testapp_name, testapp_dir)
+  testapps = []
+  for file_dir, _, file_names in os.walk(testapp_dir):
+    if testapp_name in file_names:
+      testapps.append(os.path.join(file_dir, testapp_name))
+
+  if not testapps:
+    logging.error("No testapps found.")
+    test_validation.write_summary(testapp_dir, "Desktop tests: none found.")
+    return 1
+  logging.info("Testapps found: %s\n", "\n".join(testapps))
+
+  logging.info("Running tests...")
+  tests = [Test(testapp_path=testapp) for testapp in testapps]
+  threads = []
+  for test in tests:
+    thread = threading.Thread(target=test.run)
+    threads.append(thread)
+    thread.start()
+  for thread in threads:
+    thread.join()
+  logging.info("Finished running tests.")
+
+  return test_validation.summarize_test_results(
+      tests, test_validation.UNITY, testapp_dir)
+
+
+def _fix_path(path):
+  """Expands ~, normalizes slashes, and converts relative paths to absolute."""
+  return os.path.abspath(os.path.expanduser(path))
+
+
+@attr.s(frozen=False, eq=False)
+class Test(object):
+  """Holds data related to the testing of one testapp."""
+  testapp_path = attr.ib()
+  # This will be populated after the test completes, instead of initialization.
+  logs = attr.ib(init=False, default=None)
+
+  # This runs in a separate thread, so instead of returning values we store
+  # them as fields so they can be accessed from the main thread.
+  def run(self):
+    """Executes this testapp."""
+    log_path = self.testapp_path + ".log"
+    try:
+      os.remove(log_path)  # Remove log file from previous runs.
+    except FileNotFoundError:
+      pass
+    args = [
+        self.testapp_path, "-batchmode", "-nographics",
+        "-TestLabManager.logPath", log_path]
+    # Unity testapps do not exit when they're done, so we need more control
```

Comment:
This is about the testapp not exit when it finishes testing. And desktop testapps are not testing in FTL.

Reply:
If this is something we want, we could add this to the testapps (Unity has an application.quit()).  Ideally hidden behind a flag of some sort, since when running it by hand we likely don't want that.

Reply:
Good idea. Created a ticket.